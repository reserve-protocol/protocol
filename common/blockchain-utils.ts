import { BigNumber } from 'ethers'
import { HardhatRuntimeEnvironment } from 'hardhat/types'

// getChainId: Returns current chain Id
export const getChainId = async (hre: HardhatRuntimeEnvironment) => {
  let _chainId
  try {
    _chainId = await hre.network.provider.send('eth_chainId')
  } catch (e) {
    console.log('failed to get chainId, falling back on net_version...')
    _chainId = await hre.network.provider.send('net_version')
  }

  if (!_chainId) {
    throw new Error(`could not get chainId from network`)
  }
  if (_chainId.startsWith('0x')) {
    _chainId = BigNumber.from(_chainId).toString()
  }
  return _chainId
}
