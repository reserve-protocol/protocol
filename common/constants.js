const { BigNumber } = require("ethers");

const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';

const ONE_ETH= BigNumber.from("1000000000000000000");

const MAX_UINT256 = BigNumber.from(2).pow(256).sub(1);
const MAX_UINT16 = (2**16) - 1;

const SCALE_FACTOR = 1e18;
const BN_SCALE_FACTOR = BigNumber.from(SCALE_FACTOR.toString());

module.exports = {
    ZERO_ADDRESS,
    MAX_UINT256,
    MAX_UINT16,
    SCALE_FACTOR,
    BN_SCALE_FACTOR,
    ONE_ETH
};
