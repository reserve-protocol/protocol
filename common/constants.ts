import { ethers } from 'hardhat'
import { BigNumber } from 'ethers'

export const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000'
export const ONE_ADDRESS = '0x0000000000000000000000000000000000000001'

export const ONE_ETH = BigNumber.from('1000000000000000000')

export const MAX_UINT256 = BigNumber.from(2).pow(256).sub(1)
export const MAX_UINT192 = BigNumber.from(2).pow(192).sub(1)
export const MAX_UINT96 = BigNumber.from(2).pow(96).sub(1)
export const MAX_UINT32 = 2 ** 32 - 1
export const MAX_UINT16 = 2 ** 16 - 1

export const SCALE_DECIMALS = 18
export const SCALE_FACTOR = 10 ** SCALE_DECIMALS
export const BN_SCALE_FACTOR = BigNumber.from(SCALE_FACTOR.toString())

// @dev Must match `IAsset.CollateralStatus`.
export enum CollateralStatus {
  SOUND,
  IFFY,
  DISABLED,
}

// @dev Must match `Governance.ProposalState`.
export enum ProposalState {
  Pending,
  Active,
  Canceled,
  Defeated,
  Succeeded,
  Queued,
  Expired,
  Executed,
}

// @dev Must match `Fixed.RoundingApproach`.
export enum RoundingMode {
  FLOOR,
  ROUND,
  CEIL,
}

// @dev Must match `GnosisTrade.TradeStatus`.
export enum TradeStatus {
  NOT_STARTED,
  OPEN,
  CLOSED,
}

export const FURNACE_DEST = '0x0000000000000000000000000000000000000001'
export const STRSR_DEST = '0x0000000000000000000000000000000000000002'

export const QUEUE_START = '0x0000000000000000000000000000000000000000000000000000000000000001'

// Auth roles
export const OWNER = ethers.utils.formatBytes32String('OWNER')
export const SHORT_FREEZER = ethers.utils.formatBytes32String('SHORT_FREEZER')
export const LONG_FREEZER = ethers.utils.formatBytes32String('LONG_FREEZER')
export const PAUSER = ethers.utils.formatBytes32String('PAUSER')
