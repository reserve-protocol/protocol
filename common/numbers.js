const { BigNumber } = require("ethers");
const { BN_SCALE_FACTOR } = require("./constants");

const bn = (x) => {
    if (BigNumber.isBigNumber(x)) return x;
    const stringified = parseScientific(x.toString());
    const integer = stringified.split('.')[0];
    return BigNumber.from(integer);
};

const fp = (x) => {
    return BN_SCALE_FACTOR.mul(x);
}
 
function parseScientific(num) {
    // If the number is not in scientific notation return it as it is
    if (!/\d+\.?\d*e[+-]*\d+/i.test(num)) return num;

    // Remove the sign
    const numberSign = Math.sign(Number(num));
    num = Math.abs(Number(num)).toString();

    // Parse into coefficient and exponent
    const [coefficient, exponent] = num.toLowerCase().split('e');
    let zeros = Math.abs(Number(exponent));
    const exponentSign = Math.sign(Number(exponent));
    const [integer, decimals] = (coefficient.indexOf('.') != -1 ? coefficient : `${coefficient}.`).split('.');

    if (exponentSign === -1) {
        zeros -= integer.length;
        num =
            zeros < 0
                ? integer.slice(0, zeros) + '.' + integer.slice(zeros) + decimals
                : '0.' + '0'.repeat(zeros) + integer + decimals;
    } else {
        if (decimals) zeros -= decimals.length;
        num =
            zeros < 0
                ? integer + decimals.slice(0, zeros) + '.' + decimals.slice(zeros)
                : integer + decimals + '0'.repeat(zeros);
    }

    return numberSign < 0 ? '-' + num : num;
}

module.exports = {
    bn,
    fp
}