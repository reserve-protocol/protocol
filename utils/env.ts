import dotenv from 'dotenv'

dotenv.config()

type IEnvVars =
  | 'MAINNET_RPC_URL'
  | 'ALCHEMY_MAINNET_RPC_URL'
  | 'GOERLI_RPC_URL'
  | 'MNEMONIC'
  | 'REPORT_GAS'
  | 'FORK'
  | 'SLOW'
  | 'PROTO'
  | 'PROTO_IMPL'
  | 'ETHERSCAN_API_KEY'
  | 'NO_OPT'
  | 'MAINNET_BLOCK'
  | 'ONLY_FAST'
  | 'JOBS'

export function useEnv(key: IEnvVars | IEnvVars[], _default = ''): string {
  if (typeof key === 'string') {
    return process.env[key] ?? _default
  }
  if (typeof key === 'object') {
    for (const s of key) {
      if (process.env[s]) {
        return process.env[s]!
      }
    }
  }
  return _default
}
