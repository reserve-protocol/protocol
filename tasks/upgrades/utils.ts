import { HardhatRuntimeEnvironment } from 'hardhat/types'
import { IComponents } from '../../common/configuration'
import { IDeployments } from '../../scripts/deployment/common'
import { isValidContract } from '../../common/blockchain-utils'

export const validateDeployments = async (
  hre: HardhatRuntimeEnvironment,
  deployments: IDeployments,
  version: string
) => {
  // Check trading lib defined
  if (!deployments.tradingLib) {
    throw new Error(
      `Missing deployed TradingLib for version ${version} in network ${hre.network.name}`
    )
  } else if (!(await isValidContract(hre, deployments.tradingLib))) {
    throw new Error(`TradingLib contract not found in network ${hre.network.name}`)
  }

  // Check Main implementation is defined
  if (!deployments.implementations.main) {
    throw new Error(
      `Missing deployed Main implementation for version ${version} in network ${hre.network.name}`
    )
  } else if (!(await isValidContract(hre, deployments.implementations.main))) {
    throw new Error(`Main contract not found in network ${hre.network.name}`)
  }

  // Check all componet implementations are defined
  const componentsKeys: string[] = [
    'backingManager',
    'basketHandler',
    'broker',
    'distributor',
    'furnace',
    'rTokenTrader',
    'rsrTrader',
    'rToken',
    'stRSR',
  ]

  componentsKeys.forEach(async (keystr) => {
    if (!deployments.implementations.components[keystr as keyof IComponents]) {
      throw new Error(
        `Missing deployed ${keystr} implementation for version ${version} in network ${hre.network.name}`
      )
    } else if (
      !(await isValidContract(
        hre,
        deployments.implementations.components[keystr as keyof IComponents]
      ))
    ) {
      throw new Error(
        `Implementation ${keystr} contract not found for version ${version} in network ${hre.network.name}`
      )
    }
  })
}
