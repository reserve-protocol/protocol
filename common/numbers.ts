import { BigNumber } from 'ethers'
import { ethers } from 'hardhat'
import { BN_SCALE_FACTOR, SCALE_DECIMALS } from './constants'

export const ZERO = BigNumber.from(0)

export const bn = (x: BigNumber | number): BigNumber => {
  if (BigNumber.isBigNumber(x)) return x
  const stringified = parseScientific(x.toString())
  const integer = stringified.split('.')[0]
  return BigNumber.from(integer)
}

export const pow10 = (exponent: number): BigNumber => {
  return bn(10).pow(exponent)
}

export const fp = (x: BigNumber | number): BigNumber => {
  if (BigNumber.isBigNumber(x)) {
    return BN_SCALE_FACTOR.mul(x)
  }
  return ethers.utils.parseUnits(x.toString(), SCALE_DECIMALS)
}

export const divCeil = (x: BigNumber, y: BigNumber): BigNumber =>
  // ceil(x/y) == (x + y - 1) / y
  x.add(y).sub(1).div(y)

function parseScientific(num: string) {
  // If the number is not in scientific notation return it as it is
  if (!/\d+\.?\d*e[+-]*\d+/i.test(num)) return num

  // Remove the sign
  const numberSign = Math.sign(Number(num))
  num = Math.abs(Number(num)).toString()

  // Parse into coefficient and exponent
  const [coefficient, exponent] = num.toLowerCase().split('e')
  let zeros = Math.abs(Number(exponent))
  const exponentSign = Math.sign(Number(exponent))
  const [integer, decimals] = (coefficient.indexOf('.') != -1 ? coefficient : `${coefficient}.`).split('.')

  if (exponentSign === -1) {
    zeros -= integer.length
    num =
      zeros < 0
        ? integer.slice(0, zeros) + '.' + integer.slice(zeros) + decimals
        : '0.' + '0'.repeat(zeros) + integer + decimals
  } else {
    if (decimals) zeros -= decimals.length
    num =
      zeros < 0
        ? integer + decimals.slice(0, zeros) + '.' + decimals.slice(zeros)
        : integer + decimals + '0'.repeat(zeros)
  }

  return numberSign < 0 ? '-' + num : num
}
