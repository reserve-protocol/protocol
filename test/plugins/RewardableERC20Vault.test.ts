import { loadFixture } from '@nomicfoundation/hardhat-network-helpers'
import { expect } from 'chai'
import { Wallet, ContractFactory, BigNumber } from 'ethers'
import { ethers } from 'hardhat'
import { bn, fp } from '../../common/numbers'
import {
  ERC20MockDecimals,
  ERC20MockRewarding,
  RewardableERC20Vault,
  RewardableERC20VaultTest,
} from '../../typechain'
import { cartesianProduct } from '../utils/cases'
import { useEnv } from '#/utils/env'
import { Implementation } from '../fixtures'
import snapshotGasCost from '../utils/snapshotGasCost'

type Fixture<T> = () => Promise<T>

interface RewardableERC20VaultFixture {
  rewardableVault: RewardableERC20Vault
  rewardableAsset: ERC20MockRewarding
  rewardToken: ERC20MockDecimals
}

const getFixture = (
  assetDecimals: number,
  rewardDecimals: number
): Fixture<RewardableERC20VaultFixture> => {
  const fixture: Fixture<RewardableERC20VaultFixture> =
    async function (): Promise<RewardableERC20VaultFixture> {
      const rewardTokenFactory: ContractFactory = await ethers.getContractFactory(
        'ERC20MockDecimals'
      )
      const rewardToken = <ERC20MockDecimals>(
        await rewardTokenFactory.deploy('Reward Token', 'REWARD', rewardDecimals)
      )

      const rewardableAssetFactory: ContractFactory = await ethers.getContractFactory(
        'ERC20MockRewarding'
      )
      const rewardableAsset = <ERC20MockRewarding>(
        await rewardableAssetFactory.deploy(
          'Rewarding Test Asset',
          'rewardTEST',
          assetDecimals,
          rewardToken.address
        )
      )

      const rewardableVaultFactory: ContractFactory = await ethers.getContractFactory(
        'RewardableERC20VaultTest'
      )
      const rewardableVault = <RewardableERC20VaultTest>(
        await rewardableVaultFactory.deploy(
          rewardableAsset.address,
          'Rewarding Test Asset Vault',
          'vrewardTEST',
          rewardToken.address
        )
      )

      return {
        rewardableVault,
        rewardableAsset,
        rewardToken,
      }
    }
  return fixture
}

const toShares = (assets: BigNumber, assetDecimals: number, shareDecimals: number): BigNumber => {
  return assets.mul(bn(10).pow(shareDecimals - assetDecimals))
}

const runTests = (assetDecimals: number, rewardDecimals: number) => {
  describe('RewardableERC20Vault', () => {
    // Decimals
    let shareDecimals: number

    // Assets
    let rewardableVault: RewardableERC20Vault
    let rewardableAsset: ERC20MockRewarding
    let rewardToken: ERC20MockDecimals

    // Main
    let alice: Wallet
    let bob: Wallet

    const initBalance = fp('10000').div(bn(10).pow(18 - assetDecimals))
    const rewardAmount = fp('200').div(bn(10).pow(18 - rewardDecimals))
    let oneShare: BigNumber
    let initShares: BigNumber

    const fixture = getFixture(assetDecimals, rewardDecimals)

    before('load wallets', async () => {
      ;[alice, bob] = (await ethers.getSigners()) as unknown as Wallet[]
    })

    beforeEach(async () => {
      // Deploy fixture
      ;({ rewardableVault, rewardableAsset, rewardToken } = await loadFixture(fixture))

      await rewardableAsset.mint(alice.address, initBalance)
      await rewardableAsset.connect(alice).approve(rewardableVault.address, initBalance)
      await rewardableAsset.mint(bob.address, initBalance)
      await rewardableAsset.connect(bob).approve(rewardableVault.address, initBalance)

      shareDecimals = await rewardableVault.decimals()
      initShares = toShares(initBalance, assetDecimals, shareDecimals)
      oneShare = bn('1').mul(bn(10).pow(shareDecimals))
    })

    describe('Deployment', () => {
      it('sets the rewardableVault rewardableAsset', async () => {
        const seenAsset = await rewardableVault.asset()
        expect(seenAsset).equal(rewardableAsset.address)
      })

      it('sets the rewardableVault reward token', async () => {
        const seenRewardToken = await rewardableVault.rewardToken()
        expect(seenRewardToken).equal(rewardToken.address)
      })
    })

    describe('alice deposit, accrue, alice deposit, bob deposit', () => {
      let rewardsPerShare: BigNumber

      beforeEach(async () => {
        // alice deposit, accrue, and claim
        await rewardableVault.connect(alice).deposit(initBalance.div(4), alice.address)
        await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)
        await rewardableVault.connect(alice).deposit(initBalance.div(8), alice.address)

        // bob deposit
        await rewardableVault.connect(bob).deposit(initBalance.div(8), bob.address)

        rewardsPerShare = await rewardableVault.rewardsPerShare()
      })

      it('alice shows correct balance', async () => {
        expect(initShares.mul(3).div(8)).equal(await rewardableVault.balanceOf(alice.address))
      })

      it('alice shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(alice.address))
      })

      it('bob shows correct balance', async () => {
        expect(initShares.div(8)).equal(await rewardableVault.balanceOf(bob.address))
      })

      it('bob shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(bob.address))
      })

      it('rewardsPerShare is correct', async () => {
        // rewards / alice's deposit
        expect(rewardsPerShare).equal(rewardAmount.mul(oneShare).div(initShares.div(4)))
      })
    })

    describe('alice deposit, accrue, alice deposit, accrue, bob deposit', () => {
      let rewardsPerShare: BigNumber
      let initRewardsPerShare: BigNumber

      beforeEach(async () => {
        // alice deposit, accrue, and claim
        await rewardableVault.connect(alice).deposit(initBalance.div(4), alice.address)
        await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)
        await rewardableVault.connect(alice).deposit(initBalance.div(4), alice.address)

        initRewardsPerShare = await rewardableVault.rewardsPerShare()

        // accrue
        await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)

        // bob deposit
        await rewardableVault.connect(alice).deposit(initBalance.div(8), bob.address)

        rewardsPerShare = await rewardableVault.rewardsPerShare()
      })

      it('alice shows correct lastRewardsPerShare', async () => {
        // rewards / alice's deposit
        expect(initRewardsPerShare).equal(rewardAmount.mul(oneShare).div(initShares.div(4)))
        expect(initRewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(alice.address))
      })

      it('bob shows correct lastRewardsPerShare', async () => {
        const expectedRewardsPerShare = rewardAmount
          .mul(oneShare)
          .div(initShares.div(4))
          .add(rewardAmount.mul(oneShare).div(initShares.div(2)))
        expect(rewardsPerShare).equal(expectedRewardsPerShare)
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(bob.address))
      })
    })

    describe('alice deposit, accrue, alice withdraw', () => {
      let rewardsPerShare: BigNumber

      beforeEach(async () => {
        // alice deposit, accrue, and claim
        await rewardableVault.connect(alice).deposit(initBalance.div(4), alice.address)
        await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)
        await rewardableVault
          .connect(alice)
          .withdraw(initBalance.div(8), alice.address, alice.address)

        rewardsPerShare = await rewardableVault.rewardsPerShare()
      })

      it('alice shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(alice.address))
      })

      it('rewardsPerShare is correct', async () => {
        // rewards / alice's deposit
        expect(rewardsPerShare).equal(rewardAmount.mul(oneShare).div(initShares.div(4)))
      })
    })

    describe('alice deposit, accrue, bob deposit, alice withdraw', () => {
      let rewardsPerShare: BigNumber

      beforeEach(async () => {
        // alice deposit, accrue, and claim
        await rewardableVault.connect(alice).deposit(initBalance.div(4), alice.address)
        await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)
        await rewardableVault.connect(bob).deposit(initBalance.div(4), bob.address)
        await rewardableVault
          .connect(alice)
          .withdraw(initBalance.div(8), alice.address, alice.address)

        rewardsPerShare = await rewardableVault.rewardsPerShare()
      })

      it('alice shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(alice.address))
      })

      it('bob shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(bob.address))
      })

      it('rewardsPerShare is correct', async () => {
        // rewards / alice's deposit
        expect(rewardsPerShare).equal(rewardAmount.mul(oneShare).div(initShares.div(4)))
      })
    })

    describe('alice deposit, accrue, bob deposit, alice fully withdraw', () => {
      let rewardsPerShare: BigNumber

      beforeEach(async () => {
        // alice deposit, accrue, and claim
        await rewardableVault.connect(alice).deposit(initBalance.div(4), alice.address)
        await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)
        await rewardableVault.connect(bob).deposit(initBalance.div(4), bob.address)
        await rewardableVault
          .connect(alice)
          .withdraw(initBalance.div(4), alice.address, alice.address)

        rewardsPerShare = await rewardableVault.rewardsPerShare()
      })

      it('alice shows correct balance', async () => {
        expect(0).equal(await rewardableVault.balanceOf(alice.address))
      })

      it('alice shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(alice.address))
      })

      it('bob shows correct balance', async () => {
        expect(initShares.div(4)).equal(await rewardableVault.balanceOf(bob.address))
      })

      it('bob shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(bob.address))
      })

      it('rewardsPerShare is correct', async () => {
        // rewards / alice's deposit
        expect(rewardsPerShare).equal(rewardAmount.mul(oneShare).div(initShares.div(4)))
      })
    })

    describe('alice deposit, accrue, alice claim, bob deposit', () => {
      let rewardsPerShare: BigNumber

      beforeEach(async () => {
        // alice deposit, accrue, and claim
        await rewardableVault.connect(alice).deposit(initBalance.div(4), alice.address)
        await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)
        await rewardableVault.connect(alice).claimRewards()

        // bob deposit
        await rewardableVault.connect(bob).deposit(initBalance.div(8), bob.address)

        rewardsPerShare = await rewardableVault.rewardsPerShare()
      })

      it('alice shows correct balance', async () => {
        expect(initShares.div(4)).equal(await rewardableVault.balanceOf(alice.address))
      })

      it('alice has claimed rewards', async () => {
        expect(rewardAmount).equal(await rewardToken.balanceOf(alice.address))
      })

      it('alice shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(alice.address))
      })

      it('bob shows correct balance', async () => {
        expect(initShares.div(8)).equal(await rewardableVault.balanceOf(bob.address))
      })

      it('bob shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(bob.address))
      })

      it('rewardsPerShare is correct', async () => {
        // rewards / alice's deposit
        expect(rewardsPerShare).equal(rewardAmount.mul(oneShare).div(initShares.div(4)))
      })
    })

    describe('alice deposit, accrue, bob deposit, accrue, bob claim, alice claim', () => {
      let rewardsPerShare: BigNumber

      beforeEach(async () => {
        // alice deposit, accrue, and claim
        await rewardableVault.connect(alice).deposit(initBalance.div(4), alice.address)
        await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)

        // bob deposit
        await rewardableVault.connect(bob).deposit(initBalance.div(4), bob.address)

        // accrue
        await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)

        // claims
        await rewardableVault.connect(bob).claimRewards()
        await rewardableVault.connect(alice).claimRewards()

        rewardsPerShare = await rewardableVault.rewardsPerShare()
      })

      it('alice shows correct balance', async () => {
        expect(initShares.div(4)).equal(await rewardableVault.balanceOf(alice.address))
      })

      it('alice has claimed rewards', async () => {
        expect(rewardAmount.add(rewardAmount.div(2))).equal(
          await rewardToken.balanceOf(alice.address)
        )
      })

      it('alice shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(alice.address))
      })

      it('bob shows correct balance', async () => {
        expect(initShares.div(4)).equal(await rewardableVault.balanceOf(bob.address))
      })

      it('bob shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(bob.address))
      })

      it('bob has claimed rewards', async () => {
        expect(rewardAmount.div(2)).equal(await rewardToken.balanceOf(bob.address))
      })

      it('rewardsPerShare is correct', async () => {
        // (rewards / alice's deposit) + (rewards / (alice's deposit + bob's deposit))
        const expectedRewardsPerShare = rewardAmount
          .mul(oneShare)
          .div(initShares.div(4))
          .add(rewardAmount.mul(oneShare).div(initShares.div(2)))
        expect(rewardsPerShare).equal(expectedRewardsPerShare)
      })
    })

    describe('does not accure rewards for an account while it has no deposits', () => {
      // alice deposit, accrue, bob deposit, alice fully withdraw, accrue, alice deposit, alice claim, bob claim
      let rewardsPerShare: BigNumber

      beforeEach(async () => {
        // alice deposit, accrue, and claim
        await rewardableVault.connect(alice).deposit(initBalance.div(4), alice.address)
        // accrue
        await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)
        // bob deposit
        await rewardableVault.connect(bob).deposit(initBalance.div(4), bob.address)
        // alice withdraw all
        await rewardableVault
          .connect(alice)
          .withdraw(initBalance.div(4), alice.address, alice.address)
        // accrue
        await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)
        // alice re-deposit
        await rewardableVault.connect(alice).deposit(initBalance.div(4), alice.address)
        // both claim
        await rewardableVault.connect(bob).claimRewards()
        await rewardableVault.connect(alice).claimRewards()

        rewardsPerShare = await rewardableVault.rewardsPerShare()
      })

      it('alice shows correct balance', async () => {
        expect(initShares.div(4)).equal(await rewardableVault.balanceOf(alice.address))
      })

      it('alice shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(alice.address))
      })

      it('alice has claimed rewards', async () => {
        expect(rewardAmount).equal(await rewardToken.balanceOf(alice.address))
      })

      it('bob shows correct balance', async () => {
        expect(initShares.div(4)).equal(await rewardableVault.balanceOf(bob.address))
      })

      it('bob shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(bob.address))
      })

      it('bob has claimed rewards', async () => {
        expect(rewardAmount).equal(await rewardToken.balanceOf(bob.address))
      })

      it('rewardsPerShare is correct', async () => {
        // (rewards / alice's deposit) + (rewards / bob's deposit)
        expect(rewardsPerShare).equal(rewardAmount.mul(oneShare).div(initShares.div(4)).mul(2))
      })
    })

    describe('correctly updates rewards on transfer', () => {
      let rewardsPerShare: BigNumber

      beforeEach(async () => {
        await rewardableVault.connect(alice).deposit(initBalance.div(4), alice.address)
        await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)
        await rewardableVault.connect(bob).deposit(initBalance.div(4), bob.address)
        await rewardableVault.connect(alice).transfer(bob.address, initShares.div(4))
        await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)
        await rewardableVault.connect(alice).deposit(initBalance.div(4), alice.address)
        await rewardableVault.connect(bob).claimRewards()
        await rewardableVault.connect(alice).claimRewards()

        rewardsPerShare = await rewardableVault.rewardsPerShare()
      })

      it('alice shows correct balance', async () => {
        expect(initShares.div(4)).equal(await rewardableVault.balanceOf(alice.address))
      })

      it('alice shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(alice.address))
      })

      it('alice has claimed rewards', async () => {
        expect(rewardAmount).equal(await rewardToken.balanceOf(alice.address))
      })

      it('bob shows correct balance', async () => {
        expect(initShares.div(2)).equal(await rewardableVault.balanceOf(bob.address))
      })

      it('bob shows correct lastRewardsPerShare', async () => {
        expect(rewardsPerShare).equal(await rewardableVault.lastRewardsPerShare(bob.address))
      })

      it('bob has claimed rewards', async () => {
        expect(rewardAmount).equal(await rewardToken.balanceOf(bob.address))
      })

      it('rewardsPerShare is correct', async () => {
        // (rewards / alice's deposit) + (rewards / (alice's deposit + bob's deposit))
        expect(rewardsPerShare).equal(
          rewardAmount
            .mul(oneShare)
            .div(initShares.div(4))
            .add(rewardAmount.mul(oneShare).div(initShares.div(2)))
        )
      })
    })
  })
}

const decimalSeeds = [6, 8, 18]
const cases = cartesianProduct(decimalSeeds, decimalSeeds)
// const cases = [[6, 6]]
cases.forEach((params) => {
  describe(`rewardableAsset assetDecimals: ${params[0]} / reward assetDecimals: ${params[1]}`, () => {
    runTests(params[0], params[1])
  })
})

export const IMPLEMENTATION: Implementation =
  useEnv('PROTO_IMPL') == Implementation.P1.toString() ? Implementation.P1 : Implementation.P0

const describeGas =
  IMPLEMENTATION == Implementation.P1 && useEnv('REPORT_GAS') ? describe.only : describe.skip

describeGas('Gas Reporting', () => {
  // Assets
  let rewardableVault: RewardableERC20Vault
  let rewardableAsset: ERC20MockRewarding

  // Main
  let alice: Wallet

  const initBalance = fp('10000')
  const rewardAmount = fp('200')

  const fixture = getFixture(18, 18)

  before('load wallets', async () => {
    ;[alice] = (await ethers.getSigners()) as unknown as Wallet[]
  })

  beforeEach(async () => {
    // Deploy fixture
    ;({ rewardableVault, rewardableAsset } = await loadFixture(fixture))

    await rewardableAsset.mint(alice.address, initBalance)
    await rewardableAsset.connect(alice).approve(rewardableVault.address, initBalance)
  })

  describe('RewardableERC20Vault', () => {
    it('deposit', async function () {
      // Deposit
      await snapshotGasCost(
        rewardableVault.connect(alice).deposit(initBalance.div(4), alice.address)
      )

      // Deposit again
      await snapshotGasCost(
        rewardableVault.connect(alice).deposit(initBalance.div(4), alice.address)
      )
    })

    it('withdraw', async function () {
      await rewardableVault.connect(alice).deposit(initBalance, alice.address)

      await snapshotGasCost(
        rewardableVault.connect(alice).withdraw(initBalance.div(2), alice.address, alice.address)
      )

      await snapshotGasCost(
        rewardableVault.connect(alice).withdraw(initBalance.div(2), alice.address, alice.address)
      )
    })

    it('claimRewards', async function () {
      await rewardableVault.connect(alice).deposit(initBalance, alice.address)

      await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)

      await snapshotGasCost(rewardableVault.connect(alice).claimRewards())

      await rewardableAsset.accrueRewards(rewardAmount, rewardableVault.address)

      await snapshotGasCost(rewardableVault.connect(alice).claimRewards())
    })
  })
})
