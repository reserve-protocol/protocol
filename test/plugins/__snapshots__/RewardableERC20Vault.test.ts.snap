// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gas Reporting RewardableERC4626Vault claimRewards 1`] = `159832`;

exports[`Gas Reporting RewardableERC4626Vault claimRewards 2`] = `83066`;

exports[`Gas Reporting RewardableERC4626Vault deposit 1`] = `119033`;

exports[`Gas Reporting RewardableERC4626Vault deposit 2`] = `85387`;

exports[`Gas Reporting RewardableERC4626Vault withdraw 1`] = `98068`;

exports[`Gas Reporting RewardableERC4626Vault withdraw 2`] = `66568`;
